### PR TITLE
Support OAuth2 authentication

### LINKED ISSUE #60: Support authentication using OAuth2
## New or Affected Resource(s)

- Just Provider Configuration

## Potential Terraform Configuration

```hcl
provider "pulsar" {
  web_service_url = "http://localhost:8080"
  key_file_path = "/path/to/privateKey"
  issuer_url = "https://dev-kt-aa9ne.us.auth0.com"
  audience = "https://dev-kt-aa9ne.us.auth0.com/api/v2/"
  client_id = "0Xx...Yyxeny",
}
```

## References

- https://pulsar.apache.org/docs/security-oauth2

### PR DESCRIPTION
**Pull request by maxsxu**
Fixes #60 

### Motivation

Support authentication using OAuth2

### Modifications

*Describe the modifications you've done.*

### Verifying this change

- [x] Make sure that the change passes the CI checks.

This change added tests and can be verified as follows:

```hcl
provider "pulsar" {
  web_service_url = "https://***.test.aws.sn2.dev"
  issuer_url = "https://***..dev"
  client_id = "VNCbYg1G37x4AK***"
  audience = "urn:sn:***"
  key_file_path = "file:///admin.json"
}
```

### Documentation

- [x] `no-need-doc` 
  docs have been updated in this PR.

### CODE REVIEW

File: pkg/admin/admin_config.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+	"github.com/streamnative/terraform-provider-pulsar/pkg/authentication"
+)
+
+type PulsarAdminConfig struct {
+	Config *common.Config
+}
+
+func (p *PulsarAdminConfig) AuthenticationType() authentication.AuthenticationType {
+	if len(p.Config.IssuerEndpoint) > 0 && len(p.Config.ClientID) > 0 &&
+		len(p.Config.Audience) > 0 && len(p.Config.KeyFile) > 0 {
```
Comment (tuteng) [inline pkg/admin/admin_config.go:32]: This field is optional, do we need to check it? for example on azure AD OAuth2 this field is not needed

File: README.md [hunk 2]
```diff
@@ -78,6 +78,10 @@ provider "pulsar" {
 | `tls_trust_certs_file_path`     | Path to a custom trusted TLS certificate file                                                                          | No       |
 | `tls_allow_insecure_connection` | Boolean flag to accept untrusted TLS certificates                                                                      | No       |
 | `api_version`                   | Used to request Apache Pulsar API service, default by 0, which represents use default version                          | No       |      
+| `audience`                      | The OAuth 2.0 resource server identifier for the Pulsar cluster                                                        | No       |
```
Comment (tuteng) [inline README.md:81]: Maybe we need to consider adding another field `scope` in order to support azure AD OAuth2 https://github.com/streamnative/pulsarctl/pull/465

File: pkg/admin/admin.go [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+	"github.com/streamnative/pulsarctl/pkg/pulsar"
+
+	"github.com/streamnative/terraform-provider-pulsar/pkg/authentication"
+)
+
+func NewPulsarAdminClient(c *PulsarAdminConfig) (pulsar.Client, error) {
+	if c.AuthenticationType() == authentication.AuthenticationOauth2 {
+		oauth2Provider, err := auth.NewAuthenticationOAuth2WithDefaultFlow(oauth2.Issuer{
+			IssuerEndpoint: c.Config.IssuerEndpoint,
+			ClientID:       c.Config.ClientID,
```
Comment (tuteng) [inline pkg/admin/admin.go:34]: Maybe we need to consider adding another field `scope` in order to support azure AD OAuth2

**Comment by tuteng** [comment]
There may be another scenario to consider, I'm not sure if the golang oauth2 library currently supports it, some users may not be able to mount the file easily, so they may want to encode the client-id and client-secret using base64, and then use it. Anyway, we can implement it in the future @maxsxu

**Comment by maxsxu** [comment]
The `scope` is added, but supporting Azure AD OAuth2 lacks features of the upstream library, which will be addressed in [pulsarctl](https://github.com/streamnative/pulsarctl) in the first place.

**Review by tuteng — APPROVED** [review]

**Pull request merged**